File: 2018/solutions/prob09.cpp
#include <iostream>
#include <string>
#include <cmath>

enum NumberType {
  NEITHER = 0,
  BOTH,
  PERFECT_SQUARE,
  PERFECT_CUBE
};

const std::string NUMBER_TYPES_MAP[] {
  "NEITHER",
  "BOTH",
  "PERFECT SQUARE",
  "PERFECT CUBE"
};

bool isPerfectSquare(double number);
bool isPerfectCube(double number);

int main() {
  double input;
  bool isPerfSquare, isPerfCube;
  enum NumberType type = NEITHER;

  while (std::cin >> input) {
    isPerfSquare = isPerfectSquare(input);
    isPerfCube = isPerfectCube(input);

    if (isPerfSquare && isPerfCube) {
      type = BOTH;
    } else if (isPerfSquare) {
      type = PERFECT_SQUARE;
    } else if (isPerfCube) {
      type = PERFECT_CUBE;
    } else {
      type = NEITHER;
    }

    std::cout << NUMBER_TYPES_MAP[type] << std::endl;
  }

  return EXIT_SUCCESS;
}

bool isPerfectSquare(double number)
{
  double squareRoot = sqrt(number);
  return ((squareRoot - floor(squareRoot)) == 0);
}

bool isPerfectCube(double number)
{
  double cubeRoot = cbrt(number);
  return ((cubeRoot - floor(cubeRoot)) == 0);
}

File: 2018/solutions/prob09.java
import java.io.*;

public class prob09 {
    public static void main(String[] args) throws IOException {
        BufferedReader reader
                = new BufferedReader(new InputStreamReader(System.in));

        String line = null;
        while ((line = reader.readLine()) != null) {
            int number = Integer.parseInt(line);

            System.out.printf("%s%n", check(number));
        }
    }

    public static String check(int number) {
        boolean isPerfectSquare = isPerfectSquare(number);
        boolean isPerfectCube = isPerfectCube(number);
        String type = "";

        if (isPerfectSquare && isPerfectCube) {
            type = "BOTH";
        } else if (isPerfectSquare) {
            type = "PERFECT SQUARE";
        } else if (isPerfectCube) {
            type = "PERFECT CUBE";
        } else {
            type = "NEITHER";
        }

        return type;
    }

    public static boolean isPerfectSquare(int number) {
        double squareRoot = Math.sqrt(number);
        return ((squareRoot - Math.floor(squareRoot)) == 0);
    }

    public static boolean isPerfectCube(int number) {
        double cubeRoot = Math.cbrt(number);
        return ((cubeRoot - Math.floor(cubeRoot)) == 0);
    }

}
